docs/developer/decision-records/2023-12-19-token-handling-refactor/README.md

```diff
@@ -0,0 +1,218 @@
+
+### TokenGenerationService
+
+`TokenGenerationService` will be changed to remove the private key parameter from its constructor and instead have a `Supplier<PrivateKey>` passed to it (which may wrap a `KeyParserRegistry`):
+
+```java
+public interface TokenGenerationService {
+
+   Result<TokenRepresentation> generate(Supplier<PrivateKey> keySupplier, JwtDecorator... decorators);
+
+}
+```
+
+### TokenValidationService
+
+Many rules that apply to token validation depend on the token type. For example, OAuth token validation requires different rules than self-issued token validation. Validation rules will be associated with a token type and resolved per validation operation (there will also be default rules that apply to all types).  The `TokenValidationService` and `TokenValidationRulesRegistry` will be refactored to accommodate these changes.
+
+`TokenValidationService` will have a `PublicKeyResolver` and `TokenValidationRules` passed when performing validation, rather than `TokenValidationServiceImpl` taking a reference to a `PublicKeyResolver` and the `TokenValidationRulesRegistry` in its constructor:
+
+```java
+public interface TokenValidationService {
+
+  Result<ClaimToken> validate(TokenRepresentation representation, PublicKeyResolver resolver, TokenValidationRule... rules);
+
+  Result<ClaimToken> validate(@NotNull String token, PublicKeyResolver resolver, TokenValidationRule... rules);
+  
+}
+```
+
+The client that invokes the `TokenValidationService` will be responsible for providing the public key resolver and assembling the rules for the current security context. Public key resolvers will implement different strategiess for returning the key:
+- IATP - The implementation will resolve key material from a DID
+- OAuth - The implementation will use the `IdentityProviderKeyResolver`
+- Consumer pull data planes - May resolve from config, a vault, or some other form of secure storage
+
+> Note that the token type cannot be passed to the `TokenValidationService` in place of the array of `TokenValidationRules` since some rules depend on the current security context. For example, audience validation requires access to an audience value tied to the current runtime request.  These rules will typically be lambdas passed to the service.
+
+### `JwtDecoratorRegistry` and `TokenValidationRulesRegistry` 
+
+The `JwtDecoratorRegistry` and `TokenValidationRulesRegistry` specializations (subclasses) will be removed and replaced with the concept of a token type, e.g. OAuth2:
+
+```java
+public interface JwtDecoratorRegistry {
```

Comment:
we should do the naming consistently: have a `TokenDecoratorRegistry` and `TokenDecorator`, and JWT-impl-classes.

```suggestion
public interface TokenDecoratorRegistry {
```